File: python_package/custom_feature_package/broyden.py
import numpy as np

def transpose(x):
    x = np.transpose([x])
    return x


def byoyden(xt, yt, gamma, case):
    N = np.size(xt, 0)
    ut = np.diff(xt, axis=0)
    ut = np.vstack((np.zeros((1, np.size(xt, 1))) + 0.01, ut))
    dt = np.diff(yt, axis=0)
    dt = np.vstack((np.zeros((1, np.size(yt, 1))) + 0.01, dt))
    At0 = np.random.rand(np.size(yt, 1),  np.size(xt, 1))

    if case == 1:
        for i in range(N):
            At = At0 + gamma * (transpose(dt[i, :]) - At0.dot(transpose(ut[i, :]))).dot([ut[i, :]]) / np.square(np.linalg.norm([ut[i, :]], ord=2))
            At0 = At
        return At






## test code
N = 300
A = np.random.rand(3, 4) * 2
xt = None
yt = None
for i in range(N):
    x = np.random.rand(1, 4)
    y = A.dot(x.transpose())

    if xt is None:
        xt = x
    else:
        xt = np.vstack((xt, x))

    if yt is None:
        yt = y.transpose()
    else:
        yt = np.vstack((yt, y.transpose()))

At = byoyden(xt, yt, 0.2, 1)
error = A - At
print(error)

File: matlab/algorithm/broyden/Broyden.m
function [At,dx,dy] = Broyden(x,y,gamma,choice)
%% 
N = size(x,1);
%%
dx = diff(x);
dx = [zeros(1,size(x,2)) + 0.01;dx];
dy = diff(y);
dy = [zeros(1,size(y,2)) + 0.01;dy];
At0 = rand(size(y,2),size(x,2));
%%
switch choice
    case 1
        %% R1       
        for i=1:N
            At = At0 + gamma * (dy(i,:)' - At0 * dx(i,:)') * dx(i,:)/(dx(i,:) * dx(i,:)');
            At0 = At;
        end
    case 2
        %% SR1
        for i=1:N
            At = At0 + (dy(i,:)' - At0 * dx(i,:)') * (dy(i,:)' - At0 * dx(i,:)')' *...
                 pinv(dx(i,:)' * (dy(i,:)' - At0 * dx(i,:)')');
            At0 = At;
        end
    case 3
        %% BFGS
        for i=1:N
            At = At0 - At0 * dx(i,:)' * dx(i,:) * At0' * pinv(dx(i,:)' * dx(i,:) * At0') + dy(i,:)' * dy(i,:) * pinv(dx(i,:)' * dy(i,:));
            At0 = At;
        end
    case 4
        %% DFP
        for i=1:N
            At = At0 + ((dy(i,:)' - At0 * dx(i,:)')*dy(i,:) + dy(i,:)' * (dy(i,:)' - At0 * dx(i,:)')') * pinv(dx(i,:)' * dy(i,:))...
                 - ((dy(i,:)' - At0 * dx(i,:)') * dx(i,:)) * (dy(i,:) * dy(i,:)') / (norm(dx(i,:)' * dy(i,:)) + 0.01);
            At0 = At;
        end
    case 5
        %% PSB
        for i=1:N
            At = At0 + ((dy(i,:)' - At0 * dx(i,:)')*dx(i,:) * 1) / (dx(i,:) * dx(i,:)')...
                - ((dy(i,:)' - At0 * dx(i,:)') * dx(i,:) * dx(i,:)' * dx(i,:)) / (dx(i,:) * dx(i,:)');
            At0 = At;
        end
        
end
end